### DISCUSSION TITLE
Support for Other LLM Providers

### DISCUSSION THREAD

**Discussion by Manish06097** [567-labs/instructor#138] [General]
Is their any other way to use this for the other LLM like Mistral , llama2...

**Comment by MrJarnould** [comment]
@Manish06097 currently, Instructor only works with OpenAI Function Calling. Have a look at https://github.com/PrefectHQ/marvin since Marvin works with other LLMs – this is @jxnl's recommendation.

**Reply by jxnl** [reply]
I agree. I want to mostly only support OpenAI function calling since it means that the library is not responsible for building the structured output prompts. There's so many ways to do it that I think a framework that forces you to use their prompts ends up being limiting.

I think [Marvin](https://github.com/PrefectHQ/marvin) is great, so is Guardrails and Outlines, you can always just `Model(**dict)` to get the Pydantic object out again.

It's definitely up for debate but my goal is to give users a great experience and focus on one provider, since I'll be building more tooling around fine-tuning function calling specifically.

When other models are being released at half-functional capabilities, I'll definitely start including them more. To do that is to probably support an OpenAI API, and then changing the base URL.

**Comment by j4k0bk** [comment]
Bump! Would love to be able to run a local Llama 2, so I avoid the cost of OpenAI, inference APIs, etc.

**Reply by jxnl** [reply]
Check out [Marvin](https://github.com/PrefectHQ/marvin) for now, I think i'd support it if there was actually some function calling support out of some of these models

**Comment by kinthaiofficial** [comment]
Provider support breadth matters a lot once you're routing across multiple tiers in a multi-agent system.

The pattern we've settled on in KinthAI: structured output extraction via Instructor sits at the junction between the orchestration layer (which decides what to extract) and the model tier (which determines which provider to call). Having Instructor abstract over providers lets you swap the underlying model without changing the extraction logic.

A few things that came up when we extended provider support:

**Cost-aware provider selection** — for structured extraction tasks, the output schema complexity often determines which provider is worth using. Simple enum extraction → Haiku-class is fine and 4-5x cheaper. Complex nested schema → needs a capable model. We ended up adding cost metadata to the Instructor call so the orchestrator can decide provider tier based on schema complexity, not just task type.

**Partial extraction fallback** — when a provider doesn't reliably produce valid JSON for a complex schema, you want to gracefully fall back to partial extraction rather than failing the whole call. Different providers have very different failure modes for structured outputs.

**Retry budget** — Instructor's retry logic by default is unbounded (until `max_retries`). In multi-agent systems where the calling agent has a cost ceiling, the retry budget needs to come from the agent's remaining allowance, not a fixed retry count.

The provider-agnostic extraction layer is also useful for building agent tools that expose structured interfaces: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons has some patterns around agent capability declarations.

Are there specific providers where the structured output support is most requested right now?
